### PR TITLE
Add support for deleted vectors in segments

### PR DESCRIPTION
**Pull request by timvisee**
Part of <https://github.com/qdrant/qdrant/issues/1045>.

This implements logic in our storage bits to support optional named vectors in the future. I set a goal to keep things performant and relatively simple.

In a later PR I will expose this functionality in the external API so this can be used.

In short:
- Based upon earlier deleted flag bitvec implementation, but heavily improved
- Add bitvec with deleted flags in all storage types
- Use `BitSlice` everywhere to determine point/vector deletions, this unifies the deleted interface across all storage types, making stuff simpler
- In mmap segment, directly map `BitSlice` on top of a mmap file. So we don't need to load/store bitvecs, and we don't need multiple bitvecs, keeping things simple
- Add deleted vector support in raw scorer (regular and quantized)
- Add deleted vector support in HNSW graphs
- Add tests to cover deleted vectors in all storage types

Some performance tweaks:
- When using mmap storage with quantization, lock deleted flags backed by mmap file in physical RAM. This prevents page faults and keeps things fast (not supported on Windows)
- Advice kernel to load delete flags backed by mmap file early with `MADV_WILLNEED` for fast subsequent access (not supported on Windows)
- Use bitvec for deleted flags, rather than a byte per vector like before
- Reworked various internal bits to make some instantiation/loops more efficient

Follow-up steps (in other PRs):
- Make query planner aware of deleted vectors, optimize search for sparse segments
- Update vacuum optimizer to rebuild sparse segments
- Expose this functionality into internal/external API

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you lint your code locally using ``cargo fmt`` command prior to submission?
3. [x] Have you checked your code using ```cargo clippy``` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/segment/src/vector_storage/memmap_vector_storage.rs [hunk 1]
```diff
@@ -85,31 +89,44 @@ impl VectorStorage for MemmapVectorStorage {
...
+            &self.deleted_path,
+            dim,
+        )?);
+
+        // Flush deleted flags into store
+        // We must do that in the updated store, and cannot do it in the previous loop. That is
+        // because the file backing delete storage must be resized, and for that we'd need to know
+        // the exact number of vectors beforehand. When opening the store it is done automatically.
+        let store = self.mmap_store.as_mut().unwrap();
+        for id in deleted_ids {
```
Comment (timvisee) [inline lib/segment/src/vector_storage/memmap_vector_storage.rs:125]: When updating mmap storage from another, we first insert new vectors in a loop, then we process all deletions in a second loop. I find this ugly, but I don't see a way to do this more elegantly. Is it good enough?

File: lib/segment/src/vector_storage/mmap_vectors.rs [hunk 2]
```diff
@@ -118,4 +140,138 @@ impl MmapVectors {
...
+/// Create a convenient [`BitSlice`] view over a [`MmapMut`].
+///
+/// This works because the internal memory mapped slice is pinned and doesn't move in memory.
+///
+/// This is unsafe because we create a shared mutable refrence with lifetime `'a`.
+///
+/// - The mmap and BitSlice should never be mutated together.
+/// - The bitslice reference should never outlive the mmap.
+/// - The caller is responsible for handling this with care.
+unsafe fn mmap_to_bitslice<'a>(mmap: &mut MmapMut) -> &'a mut BitSlice {
```
Comment (timvisee) [inline lib/segment/src/vector_storage/mmap_vectors.rs:258]: @IvanPleshkov @generall Maybe look at this in a bit more detail to make sure it is sound.

**Review by IvanPleshkov — APPROVED** [review]

**Review by generall — CHANGES_REQUESTED** [review]
- flush for mmap storage
- other changes as discussed in a call

**Comment by timvisee** [comment]
I've implemented the discussed changes.

See these commits: 566153b3d7b011231cbb91f934a962c6493a47fe...bf6e3e2a5fa2310997768831a088654919a031ef

File: lib/segment/src/vector_storage/memmap_vector_storage.rs (hunk 1 above)
Comment (timvisee) [reply lib/segment/src/vector_storage/memmap_vector_storage.rs:125]: We discussed this in a call; this is okay.

**Review by generall — APPROVED** [review]

**Pull request merged**